File: src/problem.cpp
#include "problem.hpp"

problem::problem(){
	this->env = IloEnv();
	this->stagenum = 1;
	this->stagemin = 1;
	this->stagemax = 1;
	this->colnum = 0;
	this->finalstage = 0;
	this->gpcnum = 0;
	this->gpcincolnum = 0;
	this->gpclimit = 0;
	this->collimit = 0;
	this->gpcinwidth = IloNumArray(this->env);
	this->gpcin = IloNumMatrix(this->env);
	this->gpcout = IloNumArray(this->env);
	this->gpccost = IloNumArray(this->env);
	this->initialstage = IloNumArray(this->env);
}

problem::problem(std::istream& ist) {
	this->env = IloEnv();
	this->gpcinwidth = IloNumArray(this->env);
	this->gpcin = IloNumMatrix(this->env);
	this->gpcout = IloNumArray(this->env);
	this->gpccost = IloNumArray(this->env);
	this->initialstage = IloNumArray(this->env);
	ist >> this->stagemin
		>> this->stagemax
		>> this->colnum
		>> this->finalstage
		>> this->gpcnum
		>> this->gpcincolnum
		>> this->gpcinwidth
		>> this->gpcin
		>> this->gpcout
		>> this->gpccost
		>> this->initialstage
		>> this->gpclimit
		>> this->collimit;
	this->stagenum = this->stagemin;
}

problem::problem(const problem& prob) {
	this->env = prob.env;
	this->stagemin = prob.stagemin;
	this->stagemax = prob.stagemax;
	this->stagenum = prob.stagenum;
	this->colnum = prob.colnum;	
	this->finalstage = prob.finalstage;
	this->gpcnum = prob.gpcnum;
	this->gpcincolnum = prob.gpcincolnum;
	this->gpcinwidth = prob.gpcinwidth;
	this->gpcin = prob.gpcin;
	this->gpcout = prob.gpcout;
	this->gpccost = prob.gpccost;
	this->initialstage = prob.initialstage;
	this->gpclimit = prob.gpclimit;
	this->collimit = prob.collimit;
}

void problem::dump(std::ostream& ost) {
	std::cerr << "stagemin:       " ; ost << this->stagemin       << std::endl;
	std::cerr << "stagemax:       " ; ost << this->stagemax       << std::endl;
	std::cerr << "colnum:         " ; ost << this->colnum         << std::endl;
	std::cerr << "finalstage:     " ; ost << this->finalstage     << std::endl;
	std::cerr << "gpcnum:         " ; ost << this->gpcnum         << std::endl;
	std::cerr << "gpcincolnum:    " ; ost << this->gpcincolnum    << std::endl;
	std::cerr << "gpcinwidth:     " ; ost << this->gpcinwidth     << std::endl;
	std::cerr << "gpcin:          " ; ost << this->gpcin          << std::endl;
	std::cerr << "gpcout:         " ; ost << this->gpcout         << std::endl;
	std::cerr << "goccost:        " ; ost << this->gpccost        << std::endl;
	std::cerr << "initialstage:   " ; ost << this->initialstage   << std::endl;
	std::cerr << "gpclimit:       " ; ost << this->gpclimit       << std::endl;
	std::cerr << "collimit:       " ; ost << this->collimit       << std::endl;
}

File: problem.py
#!/usr/bin/env python3

import sys
import math
import parsegpc

class problem:
    def __init__(self):
        self.order = [
            'stagemin',
            'stagemax',
            'colnum',
            'finalstage',
            'gpcnum',
            'gpcincolnum',
            'gpcinwidth',
            'gpcin',
            'gpcout',
            'gpccost',
            'initialstage',
            'gpclimit',
            'collimit',
        ]
        self.size = 0
        self.gpcs = []
        self.stagemin = 2
        self.stagemax = 2
        self.colnum = 0
        self.finalstage = 2
        self.gpcnum = 0
        self.gpcincolnum = 0
        self.gpcinwidth = []
        self.gpcin = []
        self.gpcout = []
        self.gpccost = []
        self.initialstage = []
        self.gpclimit = 0
        self.collimit = 0

    def dump(self, ost=sys.stdout):
        for content in self.order:
            val = getattr(self, content, None)
            if val != None:
                print(f'{content}: ', file=sys.stderr, end='')
                sys.stderr.flush()
                print(f'{val}', file=ost)
                ost.flush()
            else:
                print(f'{content} is not defined.', file=sys.stderr)
                sys.exit(1)

    def parse(self, raw):
        lines = raw.split('\n')
        values = ''
        for line in lines:
            if line == '':
                continue
            values += line.replace(' ', '')+','
        values = values.replace(',,', ',')
        values = values[:-1]
        variables = ','.join([f'self.{content}' for content in self.order])
        expr = variables + '=' + values
        exec(expr)
        self.gpcs = self.gengpcs()

    def gengpcs(self):
        gpcs = []
        for i in range(self.gpcnum):
            src = self.gpcin[i]
            dst = self.gpcout[i]
            cost = self.gpccost[i]
            gpcs.append({'src':src, 'dst':dst, 'cost':cost})
        return gpcs
